src/realm/sort_descriptor.cpp

```diff
@@ -591,3 +606,47 @@ void DescriptorOrdering::get_versions(const Group* group, TableVersions& version
         versions.emplace_back(table_key, group->get_table(table_key)->get_content_version());
     }
 }
+
+std::string SemanticSearchDescriptor::get_description(ConstTableRef) const
+{
+    return "KNN()";
+}
+
+void SemanticSearchDescriptor::execute(const Table& table, KeyValues& key_values) const
+{
+    if (m_k >= key_values.size()) return; // all entries already match as closest
+    
+    std::priority_queue<std::pair<float, ObjKey>> topResults;
+    
+    // Collect the k closest matches by distance
+    size_t n = std::min(m_k, key_values.size());
+    for (size_t i = 0; i < n; i++) {
+        ObjKey r = key_values.get(i);
+        float dist = table.dist_knn(m_query_data, m_column, r, m_sp);
+        topResults.push(std::pair<float, ObjKey>(dist, r));
+    }
+    float lastdist = topResults.empty() ? std::numeric_limits<float>::max() : topResults.top().first;
+    for (size_t i = m_k; i < key_values.size(); i++) {
+        ObjKey r = key_values.get(i);
+        if (!table.is_valid(r)) continue;
```

Comment:
Inefficient to check for validity first and then get object. Better to use `try_get_object` in `dist_knn` and return `std::optional<float>`